### PR TITLE
Feature/validate_account_locks_in_sanitization

### LINKED ISSUE #28293: Validate transaction account locks earlier
#### Problem

banking_stage validates transaction's account locks (by `get_account_locks()`) at late stage during bank loading accounts. It can be moved up earlier, so banking stage doesn't need to process transactions would failed on loading accounts.

#### Proposed Solution
- make `get_account_locks` part of transaction's sanitization

### PR DESCRIPTION
**Pull request by apfitzge**
#### Problem
We cannot lock accounts if we have duplicates or too many account locks. Currently, these transactions will still pass through qos and eat into CU limits. 

#### Summary of Changes
Follow up to #28292
Validate that we can actually create the account locks when we create the sanitized transaction.

Fixes #28293

### CODE REVIEW

File: sdk/src/transaction/sanitized.rs [hunk 1]
```diff
@@ -199,13 +203,8 @@ impl SanitizedTransaction {
...
         tx_account_lock_limit: usize,
     ) -> Result<TransactionAccountLocks> {
-        if self.message.has_duplicates() {
-            Err(TransactionError::AccountLoadedTwice)
-        } else if self.message.account_keys().len() > tx_account_lock_limit {
-            Err(TransactionError::TooManyAccountLocks)
-        } else {
-            Ok(self.get_account_locks_unchecked())
-        }
+        Self::validate_account_locks(self.message(), tx_account_lock_limit)?;
```
Comment (apfitzge) [inline sdk/src/transaction/sanitized.rs:206]: any thoughts on removing this since we check this when we create the `SanitizedTransaction`?

**Comment by apfitzge** [comment]
Note: I think because this moves the TooManyAccountLocks check/error into sanitization, this makes the change in #28285 useless

File: sdk/src/transaction/sanitized.rs (hunk 1 above)
Comment (taozhu-chicago) [reply sdk/src/transaction/sanitized.rs:206]: there are few other public functions can create `SanitizedTransaction`, maybe call `validate_account_locks()` everywhere?

**Comment by taozhu-chicago** [comment]
> Note: I think because this moves the TooManyAccountLocks check/error into sanitization, this makes that metric in https://github.com/solana-labs/solana/pull/28285 useless. The other added WouldExceed metrics are still useful

Would be good to keep metrics for both `AccountLoadedTwice` and `TooManyAccountLocks` tho. Maybe increment `error_counter` based on the result of `SanitizedTransaction::try_new()/try_create()`?

File: core/src/banking_stage.rs [hunk 2]
```diff
@@ -2145,6 +2146,7 @@ impl BankingStage {
                         &bank.feature_set,
                         bank.vote_only_bank(),
                         bank.as_ref(),
+                        bank.get_transaction_account_lock_limit(),
```
Comment (apfitzge) [inline core/src/banking_stage.rs:2149]: The call here will prevent us from passing `SanitizedTransaction`s that cannot be locked to qos, which is called below in `process_transactions`

File: sdk/src/transaction/sanitized.rs (hunk 1 above)
Comment (apfitzge) [reply sdk/src/transaction/sanitized.rs:206]: working on it

**Comment by apfitzge** [comment]
> Would be good to keep metrics for both AccountLoadedTwice and TooManyAccountLocks tho. Maybe increment error_counter based on the result of SanitizedTransaction::try_new()/try_create()?

Yeah, seems reasonable. I just added the tracking when we consume right now. Will push a commit shortly to accumulate for forwarded packets as well

**Pull request closed without merging**